monai/transforms/smooth_field/array.py

```diff
@@ -14,90 +14,163 @@
+        pad: number of pixels/voxels along the edges of the field to pad with `pad_val`
+        pad_val: value with which to pad field edges
         low: low value for randomized field
         high: high value for randomized field
         channels: number of channels of final output
+        spatial_size: final output size of the array, None to produce original uninterpolated field
+        mode: interpolation mode for resizing the field
+        align_corners: if True align the corners when upsampling field
+        device: Pytorch device to define field on
     """
 
     def __init__(
         self,
-        spatial_size: Union[Sequence[int], int],
-        rand_size: Union[Sequence[int], int],
-        padder: Optional[Transform] = None,
-        mode: Union[InterpolateMode, str] = InterpolateMode.AREA,
-        align_corners: Optional[bool] = None,
+        rand_size: Sequence[int],
+        pad: int = 0,
+        pad_val: float = 0,
         low: float = -1.0,
         high: float = 1.0,
         channels: int = 1,
+        spatial_size: Optional[Sequence[int]] = None,
+        mode: Union[InterpolateMode, str] = InterpolateMode.AREA,
+        align_corners: Optional[bool] = None,
+        device: Optional[torch.device] = None,
     ):
-        self.resizer: Transform = Resize(spatial_size, mode=mode, align_corners=align_corners)
-        self.rand_size: tuple = ensure_tuple(rand_size)
-        self.padder: Optional[Transform] = padder
-        self.field: Optional[np.ndarray] = None
-        self.low: float = low
-        self.high: float = high
-        self.channels: int = channels
+        self.rand_size = tuple(rand_size)
+        self.pad = pad
+        self.low = low
+        self.high = high
+        self.channels = channels
+        self.mode = mode
+        self.align_corners = align_corners
+        self.device = device
+
+        self.spatial_size: Optional[Sequence[int]] = None
+        self.spatial_zoom: Optional[Sequence[float]] = None
+
+        if low >= high:
+            raise ValueError("Value for `low` must be less than `high` otherwise field will be zeros")
+
+        self.total_rand_size = tuple(rs + self.pad * 2 for rs in self.rand_size)
+
+        self.field = torch.ones((1, self.channels) + self.total_rand_size, device=self.device) * pad_val
+
+        self.crand_size = (self.channels,) + self.rand_size
+
+        pad_slice = slice(None) if self.pad == 0 else slice(self.pad, -self.pad)
+        self.rand_slices = (0, slice(None)) + (pad_slice,) * len(self.rand_size)
+
+        self.set_spatial_size(spatial_size)
 
     def randomize(self, data: Optional[Any] = None) -> None:
-        self.field = self.R.uniform(self.low, self.high, (self.channels,) + self.rand_size)  # type: ignore
-        if self.padder is not None:
-            self.field = self.padder(self.field)
+        self.field[self.rand_slices] = torch.from_numpy(self.R.uniform(self.low, self.high, self.crand_size))
```

Comment:
Maybe use `convert_data_type` to convert to `self.device` directly?
https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/type_conversion.py#L213

Thanks.

Reply:
We should be sticking to our defined conversion routines as a matter of uniformity but I know that `uniform` will give me a numpy array that's typically very small so this conversion is going to be much faster than `convert_data_type`. One if the issues I'm trying to address with the smooth field types is improving performance and some of these compromises may be necessary.

Reply:
OK, I see your point.
Is it possible to also change the `self.field.to(self.device).clone()` part to this line?

Thanks.

Reply:
I changed it to remove `to(self.device)`, that part isn't needed but I do want to clone in case the unresized field is wanted for something and this should be fresh.

Reply:
ok, I see.

Thanks.